File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/device/JavaScriptDevice.kt
package ca.logaritm.dezel.modules.device

import android.content.Context.VIBRATOR_SERVICE
import android.content.SharedPreferences
import android.media.MediaActionSound
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import ca.logaritm.dezel.application.application
import ca.logaritm.dezel.core.JavaScriptClass
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback
import ca.logaritm.dezel.core.JavaScriptGetterCallback
import ca.logaritm.dezel.extension.fatalError
import java.util.*

/**
 * @class JavaScriptDevice
 * @since 0.4.0
 * @hidden
 */
open class JavaScriptDevice(context: JavaScriptContext): JavaScriptClass(context) {

	/**
	 * @property preferences
	 * @since 0.1.0
	 * @hidden
	 */
	private val preferences: SharedPreferences = this.context.application.getSharedPreferences("dezel.device.uuid", android.content.Context.MODE_PRIVATE)

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_name
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsGet_uuid(callback: JavaScriptGetterCallback) {

		var value = this.preferences.getString("dezel.device.uuid", null)
		if (value == null) {

			val editor = this.preferences.edit()
			if (editor == null) {
				return
			}

			value = UUID.randomUUID().toString()

			editor.putString("dezel.device.uuid", value)
			editor.apply()
		}

		callback.returns(value)
	}

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_sound
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_sound(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 1) {
			fatalError("sound requires 1 argument")
		}

		val sound = callback.argument(0)

		if (sound.isString) {

			when (sound.string) {

				"shutter" -> {
					MediaActionSound().play(MediaActionSound.SHUTTER_CLICK)
					return
				}

			}

			return
		}

		MediaActionSound().play(sound.number.toInt())
	}

	/**
	 * @method jsFunction_vibrate
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_vibrate(callback: JavaScriptFunctionCallback) {

		val vibrator: Vibrator

		try {
			vibrator = this.context.application.getSystemService(VIBRATOR_SERVICE) as Vibrator
		} catch (e: Exception) {
			return
		}

		if (callback.arguments == 0) {

			if (Build.VERSION.SDK_INT >= 26) {
				vibrator.vibrate(VibrationEffect.createOneShot(150, VibrationEffect.DEFAULT_AMPLITUDE))
			} else {
				vibrator.vibrate(150)
			}

			return
		}
	}
}

File: platforms/ios/Dezel/Dezel/Modules/Device/JavaScriptDevice.swift
import AudioToolbox

/**
 * @class JavaScriptDevice
 * @super JavaScriptClass
 * @since 0.7.0
 */
open class JavaScriptDevice: JavaScriptClass {

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_name
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsGet_uuid(callback: JavaScriptGetterCallback) {

		var uuid = UserDefaults.standard.string(forKey: "dezel.device.uuid")
		if (uuid == nil) {
			uuid = UUID().uuidString
			UserDefaults.standard.set(uuid, forKey: "dezel.device.uuid")
		}

		callback.returns(uuid!)
	}

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_sound
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsFunction_sound(callback: JavaScriptFunctionCallback) {

		if (callback.arguments < 1) {
			fatalError("sound requires 1 argument")
		}

		let sound = callback.argument(0)

		switch (sound.string) {

			case "shutter":
				AudioServicesPlaySystemSound(SystemSoundID(1108))
				return

			default:
				break
		}

		AudioServicesPlaySystemSound(SystemSoundID(sound.number.toInt()))
	}

	/**
	 * @method jsFunction_vibrate
	 * @since 0.7.0
	 * @hidden
	 */
	@objc func jsFunction_vibrate(callback: JavaScriptFunctionCallback) {

		if (callback.arguments == 0) {
			AudioServicesPlaySystemSound(SystemSoundID(4095))
			return
		}

		if (callback.arguments == 1) {

			switch (callback.argument(0).string) {

				case "peek":
					AudioServicesPlaySystemSound(SystemSoundID(1519))
					return

				case "pop":
					AudioServicesPlaySystemSound(SystemSoundID(1520))
					return

				case "cancel":
					AudioServicesPlaySystemSound(SystemSoundID(1521))
					return

				case "retry":
					AudioServicesPlaySystemSound(SystemSoundID(1102))
					return

				case "failed":
					AudioServicesPlaySystemSound(SystemSoundID(1107))
					return

				default:
					AudioServicesPlaySystemSound(SystemSoundID(4095))
					break
			}
		}

		AudioServicesPlaySystemSound(SystemSoundID(callback.argument(0).number.toInt()))
	}
}
